basic/airdrop-with-withdraw/src/lib.rs

```diff
@@ -0,0 +1,83 @@
+use scrypto::prelude::*;
+
+
+blueprint! {
+
+    struct AirdropWithWithdraw {
+        admin_badge: Address,
+        recipients: HashMap<Address, Decimal>,
+        tokens : Vault
+    }
+
+    impl AirdropWithWithdraw {
+        
+        pub fn new(token : Address) -> (Component, Bucket) {
+
+            let admin_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+                                .initial_supply_fungible(1);
+
+
+            
+            let component = Self {
+                admin_badge : admin_badge.resource_address(),
+                recipients :  HashMap::new(),
+                tokens : Vault::new(token),
+            }
+            .instantiate();
+
+            return (component, admin_badge)
+        }
+
+        fn get_required_tokens_internal(&self) -> Decimal {
+            let mut required_tokens  = Decimal::zero();
+            for val in self.recipients.values() {
+                required_tokens = required_tokens + *val ;  
+            }
+            return required_tokens;
+        }
+
+        #[auth(admin_badge)]
+        pub fn get_required_tokens(&self) -> Decimal {
+            return self.get_required_tokens_internal();
+        }
+
+        #[auth(admin_badge)]
+        pub fn add_recipient(&mut self, recipient: Address, tokens : Decimal)  {
+            assert!(!self.recipients.contains_key(&recipient) , "Recipient already exist");
+            self.recipients.insert(recipient, tokens);    
+        }
+
+        #[auth(admin_badge)]
+        pub fn put_airdrop_tokens(&mut self, tokens : Bucket)   {
+
+            let recipients_count = self.recipients.len();
+            assert!(recipients_count > 0, "You must register at least one recipient before put tokens");
+
+            let required_tokens = self.get_required_tokens_internal();
+            assert!(tokens.amount() == required_tokens, "The tokens quantity must be {}", required_tokens); 
+
+            info!("put_airdrop_tokens with recipients_count {} required_tokens {}", recipients_count,  required_tokens);
+
+            self.tokens.put(tokens); 
+        }
+
+        pub fn  get_recipient_available_tokens(&mut self, recipient_address : Address) -> Decimal
+        {
+            assert!(self.recipients.contains_key(&recipient_address), "Recipient not found");
+            let recipient_tokens  = self.recipients[&recipient_address]; 
+            info!("available_tokens : {}",recipient_tokens);
+            return recipient_tokens;
+        }
+
+        pub fn withdraw_token(&mut self, recipient_address : Address) -> Bucket
```

Comment:
After creating badges, here you can remove the `recipient_address` param and add a `#[auth(user_badge)]` macro on top of the method. You would then access the address of the badge like this in the method: `auth.resource_address()`

Reply:
Hi, thanks for the review.  This allowed me to better understand the philosophy of scrypto. 
But However I have a problem in the unit tests (try_withdraw_after_added_recipients_must_be_succeeded ) when calling the withdraw_token method . I have systematically **Insufficient balance** error. Although a deposit of user_badge was made when calling the method add_recipient on the account.  Could you tell me where my error could come from?

Reply:
I am not sure if this is what is causing the issue, but when you call `add_recipient`, the key you insert into the `recipients` map is the address of the account. Then in the `withdraw_token`, you try to find the entry in the map by specifying the address of the badge as key which does not exists.

I just realised that you will need to create a non fungible token instead of a badge for user identification, since every badges is the same so every user will be withdrawing from the same vault. Each NFT have a specific ID which can be used as the key in the `recipients` map.